### ISSUE TITLE
Uncaught Exception stops running the library

### ISSUE THREAD

**Issue by Yamakasi** [PHPSocialNetwork/phpfastcache#644]
Sorry for removing the format, it is really not helping for questions where people already investigated and know what they are running before further questions can be answered.

I'm running into something that I cannot find in the tests, examples or while searching through the library as well.

When I, with a reason for testing, change the password of a Couchbase bucket I don't get a Provider object back when I try to connect. This sounds pretty clear but the PHP errorlog shows an Uncaught Exception that I can only catch by placing a try/catch block around my:

`                $provider = CacheManager::getInstance($config['provider'], new Config([
                            'host' => $config['host'],
                            'port' => (int) $config['port'],
                            'username' => $config['user'],
                            'password' => $config['pass'],
                            'bucketName' => $config['dbname']
                                ])
                );`

So I'm wondering if the Couchbase Client SDK interacts here with the Uncaught Exception and stops further execution (see bootstrap message) of the script or something else is going on, see the log:

`[Sun Aug 26 03:05:44.286526 2018] [proxy_fcgi:error] [pid 15823] [client 192.168.255.1:52975] AH01071: Got error 'PHP message: [cb,EROR] (negotiation L:154 I:2127894937) <127.0.0.1:11210> (CTX=0x564f86ca5fe0,sasl,SASLREQ=0x564f869f9a00) Error: 0x2, Provided credentials not allowed for bucket\nPHP message: [cb,EROR] (cccp L:165 I:2127894937) <NOHOST:NOPORT> (CTX=(nil),) Could not get configuration: LCB_AUTH_ERROR (0x02)\nPHP message: [cb,EROR] (bootstrap L:170 I:2127894937) Failed to bootstrap client=0x564f86d58310. Error=LCB_AUTH_ERROR (0x02), Message=No more bootstrap providers remain\nPHP message: [cb,EROR] (pcbc/pool L:115) Failed to bootstrap LCB connection: LCB_AUTH_ERROR: Authentication failed. You may have provided an invalid username/password combination. I=0x564f86d58310\nPHP message: PHP Fatal error:  Uncaught Couchbase\\Exception: LCB_AUTH_ERROR: Authentication failed. You may have provided an invalid username/password combination in....`

I'm not able to catch the Exception otherwise then using a try/catch block around my connection part, but I would like to catch it somewhere else in my code which seems quite impossible after 2 days of debugging.

If this is Couchbase Client SDK related maybe there needs some better connection error handling available ?

**Comment by Geolim4** [comment]
Hello,

First, I think I get what happened:
You had a `Couchbase\Exception` thrown, but unfortunately you may not setup an [exception handler](http://php.net/manual/en/function.set-exception-handler.php), so php finally ended up dying miserably.

However the fact that you not managed to catch the exception manually looks weird to me.
Can you show me how you placed your try/catch block ?

Cheers,
Georges

**Comment by Yamakasi** [comment]
Hi,

I think I end up in a infinite loop in this case, so that might be the issue as well. I don't try/catch this Couchbase Exception in place but in my bootstrap of Zend Framework which catches all uncaught exceptions. 

I have tested of course with a try/catch and I can catch it, I just want to bubble it up so I can catch this one at one place at all which goes wrong and I try to debug. I can catch other Uncaught exceptions in my central place so I'm wondering if you have any idea.

(btw are you in for an idea about a feature request that can return items that will be valid in the "future" to be seen as hit ?)


Cheers,

Matt

**Comment by Geolim4** [comment]
Hello @Yamakasi,

I’m closing this issue _for now_ because of (inactivity / outdated code / …).

You can always reopen it though! :) 
Please (update the issue / add comment / clarify …).

Regards,
Georges.L

**Comment by Yamakasi** [comment]
Hi George,

I was investigating this very deeply (you want to keep the issues too clean, not needed). Do you have a clue about an uncaught exception in any way when setting the provider ?

 `Uncaught Couchbase\\Exception:`

I cannot find out where I can catch that type of error and if I need to use the phpfastcache exceptions for it.

Would be nice to know.

Thanks!

Matt

**Comment by Geolim4** [comment]
The exception can be caused mostly by an invalid credential/hostname. You may have also a firewall blocking php to communicate with couchbase on a certain port. Anyway phpfastcache should not catch this kind of exception since it's probably a misconfiguration exception rather than an unavailable php component.

**Comment by Yamakasi** [comment]
Sorry for the delay, also tested this further and I saw in a commit pretty fast (~2 weeks) after my ticket you changed something on the Exception(handling). it felt something behaved different after my composer update on the Exception handling and I saw that commit in the past. As I needed to do a lot of work on my system I thought I will check it out while doing and thinking about it at the moment I wonder: Had it anything todo with it ? It seems so.
